### PR TITLE
Integrate em optimizer

### PR DESCRIPTION
**Pull request by chbricout**
This PR add EM with a manual API using two methods:
- `em_accumulate`: to compute the update and potentially accumulate over multiple batches
- `em_step`: to perform the actual parameter update

It also provides a torch style optimizer that handles the em_accumulate automatically with a torch hook.

I added a notebook to show how to train MNIST with full EM and mini batch EM both with the manual or automatic method.
I also included tests of strict convergence when training a mixture of gaussian and a mixture of categorical. To be honest, while these test shows that at least it runs without raising error or making the loss worst, I do not think that they are very powerful tests.

This is not necessarily the final state of the PR, please comment anything you might find useful !

I also included a feature to normalize the parameters with softmax after initialization as a new parameters. It is usable with any of the current API (direct layer creation and data templates)

### CODE REVIEW

File: cirkit/backend/torch/layers/optimized.py [hunk 1]
```diff
@@ -201,6 +201,21 @@ def sample(self, x: Tensor) -> tuple[Tensor, Tensor]:
         x = torch.gather(x, dim=1, index=mixing_indices)
         return x, mixing_samples
 
+    def em_accumulate(self):
+        # The gradient automatically accumulate for sum weights
```
Comment (dylanponsford) [inline cirkit/backend/torch/layers/optimized.py]: accumulate -> accumulates

File: cirkit/backend/torch/layers/optimized.py [hunk 2]
```diff
@@ -201,6 +201,21 @@ def sample(self, x: Tensor) -> tuple[Tensor, Tensor]:
...
+        pass
+
+    def em_step(self, step_size: float, pseudocount: float, alpha: float):
+        weight = list(self.parameters())[0]
+
+        weight_grad = weight.grad.clamp(0.0)
+        unn_exp_weight = weight * weight_grad
+        exp_weight = (unn_exp_weight + pseudocount / self.weight.shape[-1]) / (
+            unn_exp_weight.sum(dim=-1, keepdim=True) + pseudocount
```
Comment (dylanponsford) [inline cirkit/backend/torch/layers/optimized.py]: I could be wrong but do we have to multiply the pseudocount by something here? Can't quite remember the formula

File: cirkit/backend/torch/em_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+class EM(Optimizer):
+    """Expectation Maximization optimizer for torch backend."""
+
+    def __init__(self, pc: TorchCircuit, lr: float, pseudocount: float, alpha: float = 1e-8):
+        """Initialize the optimizer.
+
+        Args:
+            pc: Compiled circuit to optimized.
+            lr: Learning rate / Step size.
```
Comment (dylanponsford) [inline cirkit/backend/torch/em_optimizer.py:43]: Should we call it something other than a learning rate? Maybe then it's not intuitive that it must be in [0,1]?

File: cirkit/backend/torch/em_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        if not 0.0 <= lr <= 1.0:
+            raise ValueError("lr must be in [0,1]")
+        defaults = dict(lr=lr, pseudocount=pseudocount, alpha=alpha)
+        super().__init__(pc.parameters(), defaults)
+        self.pc = pc
+
+        for layer in self.pc.layers:
+            layer.enable_em()
+
+            # Accumulate the update right after the gradient are computed
```
Comment (dylanponsford) [inline cirkit/backend/torch/em_optimizer.py]: gradient -> gradients

File: cirkit/backend/torch/layers/input.py [hunk 5]
```diff
@@ -433,6 +490,24 @@ def sample(self, num_samples: int = 1) -> Tensor:
...
+    def _sufficient_statistic(self) -> Tensor:
+        stat = torch.nn.functional.one_hot(
+            self._em_cache_input.long(), num_classes=self.num_categories
+        ).squeeze(2)
+        return stat
+
+    def _comp_em_numerator(self) -> Tensor:
+        pl = self._em_cache_output.grad.clamp(0.0)
+        return torch.einsum("ilj, ilk->ijk", pl, self._sufficient_statistic().type(pl.type()))
+
```
Comment (chbricout) [inline cirkit/backend/torch/layers/input.py:511]: Try to implement it as a primitive operation + benchmark ?

File: cirkit/backend/torch/layers/input.py [hunk 6]
```diff
@@ -304,6 +319,47 @@ def log_partition_function(self) -> Tensor:
...
+        )
+
+    def _comp_em_numerator(self) -> Tensor:
+        """Compute the numerator of the EM update.
+
+        We allow the user to override this instead of just the computation of the sufficient statistic
+        as some layer, such as the Categorical layer, can use tricks to compute the numerator with
+        less memory."""
+        suff_stat = self._sufficient_statistic()
+        pl = self._em_cache_output.grad.unsqueeze(-1)
```
Comment (chbricout) [inline cirkit/backend/torch/layers/input.py]: add .clamp(0.0)

File: cirkit/backend/torch/layers/input.py [hunk 7]
```diff
@@ -304,6 +319,47 @@ def log_partition_function(self) -> Tensor:
...
+        return torch.sum(suff_stat * pl, dim=1)
+
+    def em_accumulate(self):
+        if self._em_numerator is None:
+            self._em_numerator = self._comp_em_numerator().detach()
+        else:
+            self._em_numerator += self._comp_em_numerator().detach()
```
Comment (chbricout) [inline cirkit/backend/torch/layers/input.py]: Wrap in torch.no_grad() instead of detach()

File: cirkit/backend/torch/layers/input.py [hunk 8]
```diff
@@ -273,9 +275,22 @@ class TorchExpFamilyLayer(TorchInputFunctionLayer, ABC):
...
+    _em_numerator: Tensor | None = None
+    _em_denominator: Tensor | None = None
+
     def forward(self, x: Tensor) -> Tensor:
-        x = self.log_unnormalized_likelihood(x)
-        return self.semiring.map_from(x, LSESumSemiring)
+        out = self.log_unnormalized_likelihood(x)
+        out = self.semiring.map_from(out, LSESumSemiring)
+
+        if self._use_em and list(self.parameters())[0].requires_grad:
```
Comment (chbricout) [inline cirkit/backend/torch/layers/input.py:290]: What if someone call the model on other data (not training) before calling accumulate ?

File: cirkit/backend/torch/layers/input.py [hunk 9]
```diff
@@ -684,6 +759,17 @@ def sample(self, num_samples: int = 1) -> Tensor:
         samples = samples.permute(1, 2, 0)  # (F, K, N)
         return samples
 
+    def _sufficient_statistic(self) -> Tensor:
+        return torch.stack((self._em_cache_input, self._em_cache_input**2), dim=-1)
+
+    def em_step(self, step_size: float, pseudocount: float, alpha: float):
```
Comment (chbricout) [inline cirkit/backend/torch/layers/input.py:781]: check if requires_grad==True

File: cirkit/backend/torch/layers/input.py [hunk 10]
```diff
@@ -304,6 +319,47 @@ def log_partition_function(self) -> Tensor:
                 exponential family distributions.
         """
 
+    def _sufficient_statistic(self) -> Tensor:
+        """Compute sufficient statistics for each parameters
```
Comment (chbricout) [inline cirkit/backend/torch/layers/input.py]: parameters -> parameter

File: cirkit/backend/torch/layers/optimized.py [hunk 11]
```diff
@@ -201,6 +201,21 @@ def sample(self, x: Tensor) -> tuple[Tensor, Tensor]:
         x = torch.gather(x, dim=1, index=mixing_indices)
         return x, mixing_samples
 
+    def em_accumulate(self):
+        # The gradient automatically accumulate for sum weights
+        pass
+
+    def em_step(self, step_size: float, pseudocount: float, alpha: float):
+        weight = list(self.parameters())[0]
+
```
Comment (chbricout) [inline cirkit/backend/torch/layers/optimized.py:213]: check if requires_grad==True

File: tests/backend/torch/test_train_em.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        in_layers={  # Connections between layers (i.e. edges in the graph as an adjacency list)
+            g0: [],
+            g1: [],
+            prod: [g0, g1],
+            sl: [prod],
+        },
+        outputs=[sl],  # Nodes that are returned by the circuit
+    )
+
+
```
Comment (chbricout) [inline tests/backend/torch/test_train_em.py:66]: @dylanponsford thank you for the help with the manual step check !

File: tests/backend/torch/test_train_em.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    for epoch in range(100):
+        ll = cc(dataset)
+        loss = ll.mean()
+        loss.backward()
+        optim.step()
+        cc.zero_grad()
+
+        losses.append(-loss)
+
+    assert sorted(losses, reverse=True) == losses, "Loss should be strictly decreasing"
```
Comment (chbricout) [inline tests/backend/torch/test_train_em.py]: remove "strictly"

File: tests/backend/torch/test_train_em.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    )
+    gmm = D.MixtureSameFamily(mix, comp)
+
+    dataset = gmm.sample((1000,))
+    sc = build_gaus_symbolic_circuit(K + 1)
+    compiler = TorchCompiler(semiring="lse-sum", fold=True, optimize=True)
+    cc = compiler.compile(sc)
+    cc = cc.train()
+
+    optim = EM(cc, lr=1, pseudocount=1e-8)
```
Comment (chbricout) [inline tests/backend/torch/test_train_em.py]: pseudocount=0.0

File: cirkit/backend/torch/layers/optimized.py (hunk 1 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/optimized.py]: Done

File: cirkit/backend/torch/layers/optimized.py (hunk 2 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/optimized.py]: Discussed at the workshop, the pseudocount is divided by the number of categories in the numerator which is equivalent to multiplying the denominator.

File: cirkit/backend/torch/em_optimizer.py (hunk 3 above)
Comment (chbricout) [reply cirkit/backend/torch/em_optimizer.py:43]: Good question ! I called it "lr" to match torch API for learning rate scheduler. Basically, it looks into this `params_groups` list of dict for anything with the key "lr". We could call the parameter "step_size" in the constructor while still storing it as "lr" though but I think it might make debugging the optimizer harder.

For example, if I want to manually monitor the evolution of the step size and I don't see a "step_size" parameter, I would be a bit confused.
https://github.com/pytorch/pytorch/blob/0d62256a2b23365f8e1604297eb23a6545102aa8/torch/optim/lr_scheduler.py#L219

File: cirkit/backend/torch/em_optimizer.py (hunk 4 above)
Comment (chbricout) [reply cirkit/backend/torch/em_optimizer.py]: Done

File: cirkit/backend/torch/layers/input.py (hunk 6 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/input.py]: Done

File: cirkit/backend/torch/layers/input.py (hunk 7 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/input.py]: Done

File: cirkit/backend/torch/layers/input.py (hunk 9 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/input.py:781]: Done

File: cirkit/backend/torch/layers/optimized.py (hunk 11 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/optimized.py:213]: Done

File: tests/backend/torch/test_train_em.py (hunk 14 above)
Comment (chbricout) [reply tests/backend/torch/test_train_em.py]: DONE

File: tests/backend/torch/test_train_em.py (hunk 13 above)
Comment (chbricout) [reply tests/backend/torch/test_train_em.py]: Done

File: cirkit/backend/torch/layers/input.py (hunk 5 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/input.py:511]: After some benchmarking, there is no big difference between this and the bmm version. I decide to use bmm anyway as it is possible that the good performances of the einsum are due to the latest optimizations of torch.

File: cirkit/backend/torch/layers/input.py (hunk 10 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/input.py]: Done

File: cirkit/backend/torch/layers/input.py (hunk 8 above)
Comment (chbricout) [reply cirkit/backend/torch/layers/input.py:290]: I included a warning in the optimizer definition

**Comment by chbricout** [comment]
I have done the modifications, we just need to have a manual test for EM :eyes: (@dylanponsford)

File: tests/backend/torch/test_train_em.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+)
+from cirkit.templates import utils
+from cirkit.backend.torch.em_optimizer import EM
+
+
+def build_gaus_symbolic_circuit(units) -> Circuit:
+    # This parametrizes the mixture weights such that they add up to one.
+    weight_factory = utils.parameterization_to_factory(
+        utils.Parameterization(
+            activation="none",  # Parameterize the sum weights by using a softmax activation
```
Comment (dylanponsford) [inline tests/backend/torch/test_train_em.py]: Typo (also for the categorical circuit)

Comment (chbricout) [reply tests/backend/torch/test_train_em.py]: Fixed
